x/evm/keeper/precompiles.go

```diff
@@ -148,83 +147,77 @@ func (k *Keeper) WithPrecompiles(precompiles map[common.Address]vm.PrecompiledCo
 // the precompiles are invalid or duplicated.
 func (k *Keeper) AddEVMExtensions(ctx sdk.Context, precompiles ...vm.PrecompiledContract) error {
-	params := k.GetParams(ctx)
+	addresses := make([]common.Address, len(precompiles))
 
-	addresses := make([]string, len(precompiles))
-	precompilesMap := maps.Clone(k.precompiles)
+	precompilesMap := make(map[common.Address]vm.PrecompiledContract, len(precompiles))
+	precompilesMap = maps.Clone(k.precompiles)
 
+	// Iterate over precompiles and:
+	// - validate it doesn't already exist in the active precompiles
+	// - get the string address and add it to the active precompiles
+	params := k.GetParams(ctx)
 	for i, precompile := range precompiles {
 		// add to active precompiles
 		address := precompile.Address()
-		addresses[i] = address.String()
-
-		// add to available precompiles, but check for duplicates
-		if _, ok := precompilesMap[address]; ok {
-			return fmt.Errorf("precompile already registered: %s", address)
+		if ok := params.IsActivePrecompile(address.String()); ok {
+			return errorsmod.Wrapf(types.ErrDuplicatePrecompile, "precompile already registered: %s", address)
 		}
+
+		addresses[i] = address
 		precompilesMap[address] = precompile
 	}
 
-	params.ActivePrecompiles = append(params.ActivePrecompiles, addresses...)
-
-	// NOTE: the active precompiles are sorted and validated before setting them
-	// in the params
-	if err := k.SetParams(ctx, params); err != nil {
-		return err
+	err := k.EnablePrecompiles(ctx, addresses...)
+	if err != nil {
+		return fmt.Errorf("failed to enable precompiles: %w", err)
 	}
 
-	// update the pointer to the map with the newly added EVM Extensions
-	k.precompiles = precompilesMap
+	// set the new precompiles map in memory for faster
+	// access during the block processing
+	if !ctx.IsCheckTx() {
+		k.precompiles = precompilesMap
+	}
 	return nil
 }
 
 // IsAvailablePrecompile returns true if the given precompile address is contained in the
 // EVM keeper's available precompiles map.
-func (k Keeper) IsAvailablePrecompile(address common.Address) bool {
-	_, ok := k.precompiles[address]
-	return ok
-}
-
-// GetAvailablePrecompileAddrs returns the list of available precompile addresses.
-//
-// NOTE: uses index based approach instead of append because it's supposed to be faster.
-// Check https://stackoverflow.com/questions/21362950/getting-a-slice-of-keys-from-a-map.
-func (k Keeper) GetAvailablePrecompileAddrs() []common.Address {
-	addresses := make([]common.Address, len(k.precompiles))
-	i := 0
-
-	//#nosec G705 -- two operations in for loop here are fine
-	for address := range k.precompiles {
-		addresses[i] = address
-		i++
-	}
-
-	sort.Slice(addresses, func(i, j int) bool {
-		return bytes.Compare(addresses[i].Bytes(), addresses[j].Bytes()) == -1
-	})
-
-	return addresses
+func (k Keeper) IsAvailablePrecompile(ctx sdk.Context, address common.Address) bool {
+	params := k.GetParams(ctx)
+	return params.IsActivePrecompile(address.String())
```

Comment:
This method has now changed what's actually being returned - it's checking if an address is an active precompile, not if it's listed in the keeper map of available precompiles.

We should probably adjust the name and doc comment too